### ISSUE TITLE
Using-decls for templated member functions should create ambiguity

### ISSUE THREAD

**Issue by usx95** [llvm/llvm-project#58571]
While trying to fix #50886, I came across some inconsistency among clang, other compilers and the standard.

```cpp
struct base {
  template <int N, int M = 1>  int foo();
};
struct bar : public base {
  using base::foo;
  template <int N, int M, int P> int foo();
};
struct baz : public base {
  using base::foo;
  template <int N, int M> int foo();
};

void func() {
  bar b;
  b.foo<10, 10>(); // clang errors while others don't.
  b.foo<10, 10, 10>();

  baz bz;
  bz.foo<10, 10>(); // no one errors but base::foo and baz::foo should be ambiguous.
}
```
https://godbolt.org/z/G7af8soW8

According to my reading of standard
* `b.foo<10,10>()` should be valid as `bar::foo` should not hide `base::foo` (different template heads).
* `bz.foo<10,10>()` should ambiguous and still `baz::foo` should not hide `base::foo` (different template heads).

Reading:
* [namespace.udecl#11.sentence-1](https://eel.is/c++draft/namespace.udecl#11.sentence-1): The set of declarations named by a [using-declarator](https://eel.is/c++draft/namespace.udecl#nt:using-declarator) that inhabits a class C does not include member functions and member function templates of a base class that correspond to (and thus would conflict with) a declaration of a function or function template in C
* [basic.scope.scope#4](https://eel.is/c++draft/basic.scope.scope#4) Two declarations [correspond](https://eel.is/c++draft/basic.scope.scope#def:correspond) if they (re)introduce the same name, unless - both declare function templates with equivalent non-object-parameter-type-lists, return types (if any), [template-head](https://eel.is/c++draft/temp.pre#nt:template-head)s, and trailing [requires-clause](https://eel.is/c++draft/temp.pre#nt:requires-clause)s (if any),
* [temp.over.link#6)(https://eel.is/c++draft/temp.over.link#6): Two [template-head](https://eel.is/c++draft/temp.pre#nt:template-head)s are [equivalent](https://eel.is/c++draft/temp.over.link#def:equivalent,template-heads) if their [template-parameter-list](https://eel.is/c++draft/temp.pre#nt:template-parameter-list)s have the same length, corresponding [template-parameter](https://eel.is/c++draft/temp.param#nt:template-parameter)s are equivalent and are both declared with [type-constraint](https://eel.is/c++draft/temp.param#nt:type-constraint)s that are equivalent if either [template-parameter](https://eel.is/c++draft/temp.param#nt:template-parameter) is declared with a [type-constraint](https://eel.is/c++draft/temp.param#nt:type-constraint), and if either [template-head](https://eel.is/c++draft/temp.pre#nt:template-head) has a [requires-clause](https://eel.is/c++draft/temp.pre#nt:requires-clause), they both have [requires-clause](https://eel.is/c++draft/temp.pre#nt:requires-clause)s and the corresponding [constraint-expression](https://eel.is/c++draft/temp.constr.decl#nt:constraint-expression)s are equivalent[.](https://eel.is/c++draft/temp.over.link#6.sentence-1)

**Comment by usx95** [comment]
Recording @zygoloid comments here:
* Part of [P1787](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p1787r6.html), this was a DR, Clang’s behavior is now wrong.
* `foo<10,10>()` should be accepted and `func<10, 10>()` would be ambiguous.
* This change should be made even if this is a breaking change (previously accepted code can now be ambiguous).
CC: @erichkeane @AaronBallman @ilya-biryukov

**Comment by AaronBallman** [comment]
I agree with everything from @zygoloid with one caution about "This change should be made even if this is a breaking change (previously accepted code can now be ambiguous)." -- if it breaks too much code in practice, we might elect to not implement it as a DR but only implement it for C++23 and later.

**Comment by opensdh** [comment]
In case it helps: this is all but identical to #49583 (albeit with the *template-head* rather than the return type).
